x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/termenum/action/NodeTermEnumRequest.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.core.termenum.action;
+
+import org.elasticsearch.action.IndicesRequest;
+import org.elasticsearch.action.support.IndicesOptions;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.transport.TransportRequest;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Internal termenum request executed directly against a specific node, querying potentially many 
+ * shards in one request
+ */
+public class NodeTermEnumRequest extends TransportRequest implements IndicesRequest {
+
+    private String field;
+    private String string;
+    private long taskStartedTimeMillis;
+    private long nodeStartedTimeMillis;
+    private boolean caseInsensitive;
+    private int size;
+    private long timeout;
+    private final QueryBuilder indexFilter;
+    private Set<ShardId> shardIds;
+    private String nodeId;
+    
+
+    public NodeTermEnumRequest(StreamInput in) throws IOException {
+        super(in);
+        // Set the clock running as soon as we appear on a node.
+        nodeStartedTimeMillis = System.currentTimeMillis();
```

Comment:
it feels inconsistent to set it when deserializing, could we require callers to call `shardStartedTimeMillis` instead?

Reply:
Can we rely on nodes' clocks not drifting by the small amounts we measure here?

Reply:
I think what Adrien meant is that we shouldn't set the timer when deserializing the data. That feels fragile and inconsistent. If that represents the time when the action started on a data node then we should be able to set it once in `TransportTermsEnumAction#asyncNodeOperation` ?

Reply:
My goal was to start the timer at the earliest opportunity on the data node (ideally prior to any queuing etc) and given the NodeTermEnumRequest was passed to `TransportTermsEnumAction#asyncNodeOperation` then the deserialization would have occurred before this point?

Reply:
I don't think this is needed. We call `asyncNodeOperation` early enough to not worry too much about what happens before.